frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx

```diff
@@ -52,510 +52,441 @@ import {
-    return this.props.element.dataType === NumberInputProto.DataType.INT
+    setIsFocused(false)
   }
 
-  private isFloatData = (): boolean => {
-    return this.props.element.dataType === NumberInputProto.DataType.FLOAT
+  const onFocus = (): void => {
+    setIsFocused(true)
   }
 
-  private getMin = (): number => {
-    return this.props.element.hasMin ? this.props.element.min : -Infinity
-  }
-
-  private getMax = (): number => {
-    return this.props.element.hasMax ? this.props.element.max : +Infinity
-  }
-
-  private getStep = (): number => {
-    const { step } = this.props.element
-
-    if (step) {
-      return step
-    }
-    if (this.isIntData()) {
-      return 1
-    }
-    return 0.01
+  const updateFromProtobuf = (): void => {
+    const { value } = element
+    element.setValue = false
+    setValue(value ?? null)
+    setFormattedValue(formatValue({ value: value ?? null, ...element, step }))
+    commitValue({ value: value ?? null, source: { fromUi: false } })
   }
 
-  /** Commit state.value to the WidgetStateManager. */
-  private commitWidgetValue = (source: Source): void => {
-    const { value } = this.state
-    const { element, widgetMgr, fragmentId } = this.props
-    const data = this.props.element
-
-    const min = this.getMin()
-    const max = this.getMax()
-
-    if (notNullOrUndefined(value) && (min > value || value > max)) {
-      const node = this.inputRef.current
-      if (node) {
-        node.reportValidity()
-      }
+  // on component mount, we want to update the value from protobuf if setValue is true, otherwise commit current value
+  React.useEffect(() => {
+    const formClearHelperCopy = formClearHelper.current
+    if (element.setValue) {
+      updateFromProtobuf()
     } else {
-      const valueToBeSaved = value ?? data.default ?? null
-
-      if (this.isIntData()) {
-        widgetMgr.setIntValue(element, valueToBeSaved, source, fragmentId)
-      } else {
-        widgetMgr.setDoubleValue(element, valueToBeSaved, source, fragmentId)
-      }
+      commitValue({ value, source: { fromUi: false } })
+    }
 
-      this.setState({
-        dirty: false,
-        value: valueToBeSaved,
-        formattedValue: this.formatValue(valueToBeSaved),
-      })
+    return () => {
+      formClearHelperCopy.disconnect()
     }
-  }
 
-  /**
-   * If we're part of a clear_on_submit form, this will be called when our
-   * form is submitted. Restore our default value and update the WidgetManager.
-   */
-  private onFormCleared = (): void => {
-    this.setState(
-      (_, prevProps) => {
-        return { value: prevProps.element.default ?? null }
-      },
-      () => this.commitWidgetValue({ fromUi: true })
-    )
-  }
+    // I don't want to run this effect on every render, only on mount.
+    // Additionally, it's okay if commitValue changes, because we only call
+    // it once in the beginning anyways.
+    /* eslint-disable react-hooks/exhaustive-deps */
+  }, [])
```

Comment:
I think we have a similar `useEffect` function in the `ButtonGroup.tsx` widget [here](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/widgets/ButtonGroup/ButtonGroup.tsx#L239-L266). It might make sense to consolidate them into a standard hook we can use everywhere (can be a follow-up of this PR where we sync about together what the best way would be 🙂 ). 
The approach there is also a little bit different, where you have to pass the `widgetMgr` to the `sync...` function, so the dependencies are slightly different but on the positive side you don't have to put the `eslint-disable` check there. I don't have a strong opinion on what the better approach is.

Reply:
Yeah I like this idea of cleaning this up by creating a hook, and agree it would be easier to grok by itself in a follow up PR.